### PR TITLE
Use nearest even rounding mode for compile-time constants

### LINKED ISSUE #1671: Rounding results differ for round(0.5) vs round(x) where x=0.5f
It appears that FXC/DXC are applying a compile time optimization to numeric literals and using round-away-from-zero whereas the actual HLSL intrinsic at runtime on the GPU uses round-halves-to-nearest-even (round_ne asm). The two should be consistent, as a compile time optimization isn't intended to change the results.

    float2 main(float f : INPUT) : OUTPUT
    {
      // if input f == 0.5
      return float2(round(f), round(0.5));
      // returns float2(0.0, 1.0)
    }

`round(0.5)` -> 1
`round(1.5)` -> 2
`round(x)` -> 0 // where x = 0.5f
`round(x)` -> 2 // where x = 1.5f

(Technically the round-away-from-zero is what I actually wanted, but I'm fine with either and can work around it, so long as the two agree with each other.)

### PR DESCRIPTION
**Pull request by vcsharma**
Fixes #1671

### CODE REVIEW

File: tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp [hunk 1]
```diff
@@ -1530,7 +1531,22 @@ Value *TryEvalIntrinsic(CallInst *CI, IntrinsicOp intriOp) {
...
   } break;
   case IntrinsicOp::IOP_round: {
-    return EvalUnaryIntrinsic(CI, roundf, round);
+    // Both FXC and DXC would apply nearest even rounding mode on variables, but on compile-time constants
+    // it would apply away from zero for midway values. This would cause differences in results in cases such
+    // as round(x) (where x is 0.5) and round(0.5) where the former would evaluate to 1 and latter would
+    // evaluate to 0.
+    //
+    // For compatibility with FXC,  DXC still preserve above behavior for language version 2016 or below.
+    // However for newer language version, DXC would always use nearest even for round() intrinsic in all
```
Comment (pow2clk) [inline tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp:1540]: just trying to make sure I understand. I think "would" here should be "should"?

Comment (vcsharma) [reply tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp]: updated.

**Comment by vcsharma** [comment]
@pow2clk It would be a bit hard to tell as in how many apps (using lang version > 2016) will have dependency on this behavior, but one could always use lang version 2016 (-HV 2016) to have backward compatibility with FXC. As an aside, we have in the past supported a few other FXC back compat scenarios supported using lang version 2016. Also, in general, it seems like a bug to me that we have not been consistent with rounding behavior w.r.t. round() intrinsic. So I think it would be good to apply the rounding mode in a consistent way in newer lang versions for DXC.

**Comment by pow2clk** [comment]
That sounds reasonable. I didn't realize 2016 was an established cutoff for such things.

File: tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp [hunk 2]
```diff
@@ -1527,7 +1528,23 @@ Value *TryEvalIntrinsic(CallInst *CI, IntrinsicOp intriOp) {
     return EvalUnaryIntrinsic(CI, floorf, floor);
   } break;
   case IntrinsicOp::IOP_round: {
-    return EvalUnaryIntrinsic(CI, roundf, round);
+    // Both FXC and DXC would apply nearest even rounding mode on variables, but on compile-time constants
+    // it would apply away from zero for midway values. This would cause differences in results in cases such
+    // as round(x) (where x is 0.5) and round(0.5) where the former would evaluate to 1 and the latter would
```
Comment (tex3d) [inline tools/clang/lib/CodeGen/CGHLSLMSFinishCodeGen.cpp:1533]: cases reversed in comment (confusing)

**Pull request merged**